internal/charmstore/store_test.go

```diff
@@ -1436,3 +1516,208 @@ func (s *StoreSuite) TestAddBundleArchiveIndexed(c *gc.C) {
 	c.Assert(err, gc.IsNil)
 	s.checkAddBundle(c, bundleArchive, true)
 }
+
+var findBestEntityTests = []struct {
+	url       string
+	expectURL string
+	expectErr string
+}{{
+	url:       "~charmers/trusty/wordpress-10",
+	expectURL: "~charmers/trusty/wordpress-10",
+}, {
+	url:       "~charmers/trusty/wordpress",
+	expectURL: "~charmers/trusty/wordpress-12",
+}, {
+	url:       "trusty/wordpress-11",
+	expectURL: "~charmers/trusty/wordpress-11",
+}, {
+	url:       "trusty/wordpress",
+	expectURL: "~mickey/trusty/wordpress-13",
+}, {
+	url:       "wordpress",
+	expectURL: "~mickey/trusty/wordpress-13",
+}, {
+	url:       "~mickey/wordpress-12",
+	expectURL: "~mickey/trusty/wordpress-12",
+}, {
+	url:       "~mickey/precise/wordpress",
+	expectURL: "~mickey/precise/wordpress-12",
+}, {
+	url:       "mysql",
+	expectErr: "entity not found",
+}}
+
+func (s *StoreSuite) TestFindBestEntity(c *gc.C) {
+	store, err := NewStore(s.Session.DB("juju_test"), nil, nil)
+	c.Assert(err, gc.IsNil)
+	err = store.DB.Entities().Insert(&mongodoc.Entity{
```

Comment:
I'd add a promulgated precise wordpress here, to check wordpress still returns the trusty one, and that precise/wordpress returns the precise one.
I'd also add some tests for bundles.
Exercising the not found error would be nice too here.